proposals/Fleetctl-trigger.md

```diff
@@ -0,0 +1,84 @@
+mechanism currently to trigger async jobs on an ad hoc basis.
+
+## Proposal
+
+### New CLI command `fleetctl trigger --name <NAME>`
+- Upon this command, the CLI client makes a request to a new authenticated endpoint (see below) to
+  trigger an ad hoc run of the named schedule. 
+
+### New `schedule` option `schedule.WithTrigger` 
+- This option adds a `trigger` channel on the `schedule` struct that will trigger an ad hoc run of
+  the scheduled jobs.
+- The trigger channel for each `schedule` is exposed via a new `schedules` map on the `Service` struct. 
+
+### New endpoint `GET /trigger?name={:name}` 
+- The request handler first calls `ds.Lock` to check if the named schedule is locked.
+- If the named schedule is unlocked, request handler sends a trigger signal on the schedule's
+  trigger channel and the server responds with status `202 Accepted`.  
+- If the named schedule is locked (presumably because the schedule is currently running), the server
+  responds with status `409 Conflict` and includes a message indicating the schedule is currently
+  locked. It is up to the user to retry. To facilitate retries, the response message could be expanded to
+  include additional status information, such as the expiration time of the current lock.
+
+### Schedule locks
+- Currently, lock duration is based on the schedule interval. 
+  - Once an instance takes the lock, it will hold the lock for the duration of the interval, even
+    after it has completed the jobs in the schedule. 
+  - For long-running jobs, the lock may expire before the current instance completes its run,
+    meaning that it is currently possible for another instance to start an overlapping job.
+  - If the lockholder instance is terminated or killed, locks are not released, which may frustrate
+    a user's attempt to configure a shorter schedule interval before the lock held by the dead
+    instance expires.
+
+- Under this proposal, locks become more dynamic.
+  - Current lockholder releases its lock once it finishes running the schedule.
+  - Graceful shutdown process handles release of locks upon termination signals. Jobs are
+    preemptable and an interrupt function must be specified for each job, 
+    e.g., `schedule.New(...).WithJob("job_name", jobFn, interruptFn)`.
+  - As a fallback for cases that can't be handled via graceful shutdown (e.g., `SIGKILL`), the
+    expiration for a lock is initially set to a relatively short default duration (e.g., 5 minutes).
+    The expiration is then periodically extended by the current instance so long as scheduled jobs
+    are running.  If the current instance dies without graceful shutdown, the lock will only be held
+    by the dead instance for a short period.
+
+### Additional UX considerations 
+- What are some potential options that would be useful for the `fleetctl trigger` command?
+  - Request the current status of the the named schedule without triggering a new run.
```

Comment:
I personally find it easier if all "trigger related stuff" is within `fleetctl trigger`.

Reply:
I also think `fleetctl trigger --name NAME --status` is a good place, especially since you're likely to be running that before or after trying to trigger the job, so you just have to add the flag (maybe we should print its current status when it fails to trigger, save a step).